src/index.js

```diff
@@ -63,54 +63,51 @@ export function manipulatePluginOptions(pluginOpts) {
 }
 
-export default ({ types: t }) => ({
-  manipulateOptions(babelOptions) {
-    let findPluginOptions = babelOptions.plugins.find(plugin => plugin[0] === this)[1];
-    findPluginOptions = manipulatePluginOptions(findPluginOptions);
-
-    this.customCWD = findPluginOptions.cwd;
-  },
-
-  pre(file) {
-    let { customCWD } = this.plugin;
-    if (customCWD === 'babelrc') {
-      const startPath = (file.opts.filename === 'unknown')
-        ? './'
-        : file.opts.filename;
-
-      const { file: babelFile } = findBabelConfig.sync(startPath);
-      customCWD = babelFile
-        ? path.dirname(babelFile)
-        : null;
-    }
-
-    this.moduleResolverCWD = customCWD || process.cwd();
-  },
-
-  visitor: {
-    CallExpression: {
-      exit(nodePath, state) {
-        if (nodePath.node.seen) {
-          return;
-        }
+export default ({ types: t }) => {
+  const importVisitors = {
+    CallExpression(nodePath, state) {
+      transformRequireCall(t, nodePath, mapModule, state, this.moduleResolverCWD);
+      transformJestCalls(t, nodePath, mapModule, state, this.moduleResolverCWD);
+      transformSystemImportCall(t, nodePath, mapModule, state, this.moduleResolverCWD);
+    },
+    ImportDeclaration(nodePath, state) {
+      transformImportCall(t, nodePath, mapModule, state, this.moduleResolverCWD);
+    },
+    ExportDeclaration(nodePath, state) {
+      transformImportCall(t, nodePath, mapModule, state, this.moduleResolverCWD);
+    },
+  };
 
-        transformRequireCall(t, nodePath, mapModule, state, this.moduleResolverCWD);
-        transformJestCalls(t, nodePath, mapModule, state, this.moduleResolverCWD);
-        transformSystemImportCall(t, nodePath, mapModule, state, this.moduleResolverCWD);
+  return {
+    manipulateOptions(babelOptions) {
+      let findPluginOptions = babelOptions.plugins.find(plugin => plugin[0] === this)[1];
+      findPluginOptions = manipulatePluginOptions(findPluginOptions);
 
-        // eslint-disable-next-line no-param-reassign
-        nodePath.node.seen = true;
-      },
+      this.customCWD = findPluginOptions.cwd;
     },
-    ImportDeclaration: {
-      exit(nodePath, state) {
-        transformImportCall(t, nodePath, mapModule, state, this.moduleResolverCWD);
-      },
+
+    pre(file) {
+      let { customCWD } = this.plugin;
+      if (customCWD === 'babelrc') {
+        const startPath = (file.opts.filename === 'unknown')
+          ? './'
+          : file.opts.filename;
+
+        const { file: babelFile } = findBabelConfig.sync(startPath);
+        customCWD = babelFile
+          ? path.dirname(babelFile)
+          : null;
+      }
+
+      this.moduleResolverCWD = customCWD || process.cwd();
     },
-    ExportDeclaration: {
-      exit(nodePath, state) {
-        transformImportCall(t, nodePath, mapModule, state, this.moduleResolverCWD);
+
+    visitor: {
+      Program: {
+        exit(programPath, state) {
+          programPath.traverse(importVisitors, state);
```

Comment:
I thought of other solutions, too, but each one of them fell short:
- caching the paths won't help with all the edge cases: let's say that path A -> path B, so path B is marked as untouchable - what about path B appearing in some other place? This could result in the loss of determinism (result would depend on the order of the paresd files and would be bad anyway)
- "marking" the paths in any way is not viable - I tried wrapping the path in the `String` object, which didn't play with the other plugin; also the node may get replaced by other plugins which will result in loss of the metadata (`babel-plugin-transform-es2015-modules-commonjs` is taking `value` and putting it into a new `StringLiteral` node)

FWIW `babel-plugin-transform-es2015-modules-commonjs` is also doing something similar to `traverse` on Program exit (it doesn't have to do the full traverse, because `import` is only allowed at the top level).